### PR TITLE
Update the Asset Pipeline Guide [skip ci]

### PR DESCRIPTION
**Pull request by brenogazzola**
# Intro

Rails 7 brought with it an overhaul of the Asset Pipeline in the form of multiple new gems that either introduced a new way of handling assets (importmaps-rails), or broke down existing gems (propshaft, jsbundling-rails, cssbundling-rails) into smaller parts that make use of existing standalone tools and therefore support modern features that were previously unavailable.

Now that things are a bit more stable and developers are already upgrading (and getting confused due to amount of options), it's time to update the rails guides.

# How should we go about this?
Got some feedback and guidance from the core devs on my [original guide](https://discuss.rubyonrails.org/t/guide-to-rails-7-and-the-asset-pipeline/80851), so here's the basics...

Primary Guidelines:
- Provide a **single** default path and explain it all the way. That's import-maps.
- Avoid at all costs option proliferation. Do not give devs a bunch of options and let then figure it out.

Secondary Guidelines:
- Focus on relevant options for Rails 7+;
- Relegate legacy options (like webpacker) to footnotes that link elsewhere;
- Default 1: Import maps + Sprockets
- Default 2: Bundling + Sprockets
- Future Default: Propshaft
- Legacy Default: Webpacker

Proposed structure:
1. There are two ways to do JavaScript in Rails 7: Bundling and import maps
2. Import maps is the default. Here's how that works. 
3. It has these caveats, if those are relevant to you, go with bundling. 
4. Both approaches lean on an asset processor. 
5. The default one is Sprockets. It'll work out of the box like this. 
6. But side note, we're working on a replacement for Rails 8+ called Propshaft, and if you can live with [caveats]. 
7. Here's how Propshaft works.

# Todo
- [ ] Rewrite the base guide
- [ ] Update the Sprockets guide
- [ ] Update the Webpacker guide
- [ ] 

# If you’d like to help
For this PR, make changes, open a PR. Or post suggestions here. Or join us in the asset pipeline channel in discord.

### CODE REVIEW

File: guides/source/asset_pipeline.md [hunk 1]
```diff
@@ -7,1174 +7,39 @@ This guide covers the asset pipeline.
...
-  end
-end
-```
+### Webpacker
+Shipped with Rails 5.2, as a wrapper around the complexity of Webpack/Node/Yarn, this gem could completely replace Sprockets or simply take over javascript transpiling and bundling. It provided Rails “out of the box” support for SPA frameworks like React.
 
-Now that you have a module that modifies the input data, it's time to register
-it as a preprocessor for your mime type.
+### Import Maps
+Shipped with Rails 7.0, it replaces Sprockets as the default asset pipeline gem. Although it eliminates the need for node/yarn and other complex tooling, it requires the application using it to be deployed in an environment that supports HTTP/2, otherwise it causes severe performance problems.
```
Comment (t27duck) [inline guides/source/asset_pipeline.md:42]: It probably should be meantioned that import maps are only for Javascript and not CSS? That's been a big gotcha I've noticed from people asking about it. They are expecting import maps to also handle the CSS.

Comment (bradgessler) [reply guides/source/asset_pipeline.md:42]: I'd go one step further and tell people what they should use for CSS. Saying that importmaps is just for JS still confuses me when I try to make sense of Rails 7+ assets. The JS story is clear to me, but CSS remains unclear.

**Comment by bradgessler** [comment]
I asked a few questions in Discord that I'm distilling into some notes here.

## Sweeping generalization

I was throwing a few things out there to distill what pipelines should look like, which landed at:

For the default/basic asset usecase:

```
Rails 7: Import Maps + Sprockets
Rails 8: Import Maps + Propshaft
```

If you need advanced javascript tooling:

```
Rails 7: jsbundling + Sprockets
Rails 8: jsbundling + Propshaft 
```

"Advanced" didn't make much sense to me, so I did some more digging. Turns out "advanced" means "you need to use yarn and/or node to manage JS & CSS dependencies".

## What's the difference between `cssbundling-rails`, `jsbundling-rails`, `dartsass-rails`, etc?

After a few more questions, I learned that `cssbundling-rails` and `jsbunding-rails` gems all wrap `yarn` to manage dependencies. That wasn't obvious to me before, but now it is after asking a few questions.

I also asked "when would I use `dartsass-rails` vs the `dart` option for `cssbundling-rails`?". Again, it came back to whether or not I needed to use node/yarn to manage dependencies.

## What's the overarching goal of future Rails asset pipelines?

If I had to distill it into as few words as possible, Rails wants to ship without `node` or `yard` as a dependency. Instead it will opt for `importmaps`, which requires no compilation, and propshaft.

That said, there are more complex use-cases for front-end developers where they may need to use npm or yarn to manage front-end dependencies. In that case, `cssbundling-rails` and `jsbundling-rails` would be the recommended path forward.

## Upgrade path from the basic to advanced use case

A point I made is that switching from the basic asset management use cases to the more complex requires a lot of changes to configuration files, location of asset files, etc. How might this migration be made less complex? Location of assets, tooling, documentation, etc. could all help.

**Comment by brenogazzola** [comment]
About Prooshaft, the upgrade guide was written for apps running the sprockets/webpacker combo. This is why it goes full node/yarn tooling and mentions the bundling gems (jsbundling and cssbundling)

But Propshaft by itself does not require them. It only does fingerprint, so if you pair it with importmaps and tailwind or a BYO css library, neither are required.

File: guides/source/asset_pipeline.md [hunk 2]
```diff
@@ -3,81 +3,60 @@
 The Asset Pipeline
 ==================
 
-This guide covers the asset pipeline.
+This guide covers the asset pipeline as it was implemented by Sprockets.
```
Comment (p8) [inline guides/source/asset_pipeline.md:6]: Doesn't this describe the Import Maps and Sprockets guide?
```suggestion
This guide covers the asset pipeline using Import Maps and Sprockets.
```

Maybe also link to the Sprockets guide:
```
The Sprockets only asset pipeline guide has been moved to a seperate guide.
```

File: guides/source/asset_pipeline.md (hunk 1 above)
Comment (KonnorRogers) [reply guides/source/asset_pipeline.md]: Just wait until module type assertions come to the browser and then this is no longer true...

🙈 

https://web.dev/css-module-scripts/

**Pull request closed without merging**